Hermes/appscale/hermes/stats/stats_app.py

```diff
@@ -0,0 +1,400 @@
+from appscale.hermes.stats.producers.process_stats import ProcessesStatsSource
+from appscale.hermes.stats.producers.proxy_stats import ProxiesStatsSource
+from appscale.hermes.stats.pubsub_base import StatsPublisher
+from appscale.hermes.stats.subscribers.cache import StatsCache, \
+  ClusterStatsCache
+from appscale.hermes.stats.subscribers.profile import ClusterNodesProfileLog, \
+  ClusterProxiesProfileLog, ClusterProcessesProfileLog
+
+
+@attr.s
+class LocalStats(object):
+  """ Container for high level information related to local stats """
+  cache_size = attr.ib()
+  update_interval = attr.ib()
+  cache = attr.ib(default=None)
+  publisher = attr.ib(default=None)
+
+
+@attr.s
+class ClusterStats(object):
+  """ Container for high level information related to cluster stats """
+  cache_size = attr.ib()
+  update_interval = attr.ib()
+  cache = attr.ib(default=None)
+  publisher = attr.ib(default=None)
+  included_field_lists = attr.ib(default=None)
+
+
+@attr.s
+class HandlerInfo(object):
+  """ Container for handler information """
+  handler_class = attr.ib()
+  init_kwargs = attr.ib()
+
+
+class StatsApp(object):
+  """
+  This class holds logic related to start of stats pubsub and API endpoints.
+
+  There are 6 possible kinds of stats:
+   - local node stats (cpu, memory, disk, ...);
+   - local processes stats (cpu, memory, ... per monitored process);
+   - local proxies stats (haproxy stats collected from local haproxy socket);
+   - cluster node stats (dict[node_ip, local_node_stats]);
+   - cluster processes stats (dict[node_ip, local_processes_stats]);
+   - cluster proxies stats (dict[node_ip, local_proxies_stats]);
+
+  Statistics works differently on master node, load balancer node and other
+  regular nodes.
+
+  Additionaly to that stats supports different levels of severity:
+   - tracking processes can be disabled;
+   - csv profile log can be written optionally;
+   - cluster stats collected on master node can be reduced by disabling
+     verbose cluster stats. It tells cluster stats collector to request only
+     limited number of fields;
+  """
+
+  def __init__(self, is_master, track_processes, write_profile,
+               verbose_cluster_stats):
+    """ Initializes all properties which will be used to configure stats
+
+    Args:
+      is_master: a boolean indicating whether this node should collect statistics
+              from other nodes
+      track_processes: a boolean indicating whether process stats should be
+              collected and published
+      write_profile: a boolean indicating whether CSV log should be written
+      verbose_cluster_stats: a boolean indicating whether all available stats
+              should be collected on master node (otherwise master will
+              select only specific important fields)
+    """
+    my_ip = appscale_info.get_private_ip()
+    lb_ips = appscale_info.get_load_balancer_ips()
+
+    self._is_lb = my_ip in lb_ips
+    if is_master is not None:
```

Comment:
I didn't want to check if this node is master in two places. So this check is done only in main module.